### ISSUE TITLE
Wrong coverage in for loop `because the loop didn't complete`

### ISSUE THREAD

**Issue by MaaniBeigy** [coveragepy/coveragepy#1175]
Hi,
I have written a method inside a class:

```python
class EhrModifier:
    def __init__(self, data):
        self.data = data

    def adt_modify(
        self,
        remodel_totalothercosts: bool = True
    ):
        if remodel_totalothercosts:
            if self.data.get('composition') is not None:
                if self.data.get('composition').get('billSummary') is not None:
                    if self.data.get('composition').get('billSummary').get(
                        'totalOtherCosts'
                    ) is not None:
                        self.data['composition']['billSummary'][
                            'totalotherCosts'
                        ] = {}
                        for i in list(
                            self.data['composition']['billSummary'][
                                'totalOtherCosts'
                            ]
                        ):
                            if i.get('value') is not None:
                                if i.get('value').get('magnitude') is not None:
                                    self.data['composition']['billSummary'][
                                        'totalotherCosts'
                                    ].update({
                                        str(i['name'][
                                            'coded_string'
                                        ]).replace(".", "_"): i[
                                            'value'
                                        ]['magnitude']})
                                else:
                                    raise Exception(
                                        """EHR packet has no value.magnitude\
 in totalOtherCosts paths"""
                                    )
                            else:
                                raise Exception(
                                    """EHR packet has no value in\
 totalOtherCosts paths"""
                                )
                    else:
                        raise Exception(
                            "EHR packet has no totalOtherCosts paths"
                        )
                else:
                    raise Exception("EHR packet has no billSummary paths")
            else:
                raise Exception("EHR packet has no composition paths")

        else:
            pass
```

The test function for the success scenario is:

```python
def test_pass():
    data = {
        'composition': {
            'billSummary': {
                'totalOtherCosts': [
                    {
                        'name': {
                            'terminology_id': 'thritaEHR.otherCost',
                            'value': 'one',
                            'coded_string': '1'},
                        'value': {
                            'unit': 'Rial',
                            'magnitude': 2700.0}
                    }, {
                        'name': {
                            'terminology_id': 'thritaEHR.otherCost',
                            'value': 'two',
                            'coded_string': '2'
                        },
                        'value': {
                            'unit': 'Rial',
                            'magnitude': 2700.0
                        }
                    }
                ]

            }
        }
    }
    instance = EhrModifier(data)
    instance.adt_modify(remodel_totalothercosts=True)

    assert instance.data['composition']['billSummary']['totalotherCosts'] == {
        '1': 2700.0,
        '2': 2700.0
    }
```

The tests pass, but in coverage, I receive `line 44 didn't return from function 'adt_modify', because the loop on line 44 didn't complete`. The referred line is `for i in list(`. The result clearly shows the for loop has successfully completed, but the coverage seems to be spurious. 

**To Reproduce**
1. What version of Python are you using? `Python 3.8.5`
1. What version of coverage.py are you using? `Version: 5.5`
1. What versions of what packages do you have installed?
```
apturl==0.5.2
arrow==1.1.0
astroid==2.5.7
attrs==19.3.0
autopep8==1.5.7
bcrypt==3.1.7
beautifulsoup4==4.8.2
binaryornot==0.4.4
blinker==1.4
Brlapi==0.7.0
certifi==2019.11.28
chardet==3.0.4
Click==7.0
colorama==0.4.3
command-not-found==0.3
cookiecutter==1.7.3
coverage==5.5
coverage-badge==1.0.1
cryptography==2.8
cupshelpers==1.0
cycler==0.10.0
DateTime==4.3
dbus-python==1.2.16
decorator==4.4.2
defer==1.0.6
defusedxml==0.7.1
distro==1.4.0
distro-info===0.23ubuntu1
duplicity==0.8.12.0
entrypoints==0.3
et-xmlfile==1.1.0
fasteners==0.14.1
flake8==3.9.2
flake8-html==0.4.1
ftfy==6.0.3
future==0.18.2
GDAL==3.0.4
genbadge==1.0.0
greenlet==1.1.0
html5lib==1.0.1
httplib2==0.14.0
idna==2.8
importlib-metadata==1.5.0
iniconfig==1.1.1
ipython-genutils==0.2.0
isort==5.8.0
Jinja2==2.10.1
jinja2-time==0.2.0
jsonschema==3.2.0
jupyter-core==4.6.3
keyring==18.0.1
Khayyam==3.0.17
kiwisolver==1.0.1
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.2
lazr.uri==1.0.3
lazy-object-proxy==1.6.0
lice==0.6
lockfile==0.12.2
louis==3.12.0
lxml==4.5.0
macaroonbakery==1.3.1
mailchecker==4.0.7
Mako==1.1.0
MarkupSafe==1.1.0
matplotlib==3.1.2
mccabe==0.6.1
monotonic==1.5
more-itertools==4.2.0
nbformat==5.0.4
netifaces==0.10.4
numpy==1.17.4
oauthlib==3.1.0
olefile==0.46
openpyxl==3.0.7
OWSLib==0.19.1
packaging==20.9
pandas==1.2.4
paramiko==2.6.0
persiantools==2.1.1
pexpect==4.6.0
phonenumbers==8.12.24
Pillow==7.0.0
plotly==4.4.1
pluggy==0.13.1
poyo==0.5.0
protobuf==3.6.1
psycopg2==2.8.4
py==1.10.0
pybadges==2.2.1
pycairo==1.16.2
pycodestyle==2.7.0
pycups==1.9.73
pyflakes==2.3.1
Pygments==2.3.1
PyGObject==3.36.0
PyJWT==1.7.1
pylint==2.8.2
pymacaroons==0.13.0
pymssql==2.2.1
PyNaCl==1.3.0
pyodbc==4.0.30
pyparsing==2.4.6
pyproj==2.5.0
pypyodbc==1.3.4
PyQt5==5.14.1
pyRFC3339==1.1
pyrsistent==0.15.5
pytest==6.2.4
pytest-cov==2.12.1
python-apt==2.0.0+ubuntu0.20.4.4
python-benedict==0.24.0
python-dateutil==2.7.3
python-debian===0.1.36ubuntu1
python-fsutil==0.5.0
python-slugify==5.0.2
pytz==2019.3
pyxdg==0.26
PyYAML==5.3.1
reportlab==3.5.34
requests==2.25.1
requests-unixsocket==0.2.0
retrying==1.3.3
SecretStorage==2.3.1
simplejson==3.16.0
sip==4.19.21
six==1.14.0
soupsieve==1.9.5
SQLAlchemy==1.4.17
systemd-python==234
text-unidecode==1.3
toml==0.10.2
traitlets==4.3.3
ubuntu-advantage-tools==20.3
ubuntu-drivers-common==0.0.0
ufw==0.36
unattended-upgrades==0.1
urllib3==1.25.8
usb-creator==0.3.7
wadllib==1.3.3
wcwidth==0.2.5
webencodings==0.5.1
wrapt==1.12.1
xkit==0.0.0
xmltodict==0.12.0
xunitparser==1.3.3
zipp==1.0.0
zope.interface==5.4.0
```
1. What code are you running?  written above
1. What commands did you run? ```coverage run -m filename```

**Expected behavior**  
written above. It seems the coverage must be 100%.

**Comment by MaaniBeigy** [comment]
Although I can't understand the reason, I could solve the problem by modifying the last line `else: pass`  code to:

```python
else:
            self.data['composition']['billSummary']['totalotherCosts'] = {}
```
And writing a test as:

```python
def test_adt_modify_remodel_totalothercosts_false_pass():
    """Tests the adt_modify function where adt_modify with false
    remodel_totalothercosts passes"""
    with open(os.path.join(DIR, 'bill_sum_big.json'), 'r') as file:
        data = json.load(file)  # Loading JSON file
    instance = EhrModifier(data)
    instance.adt_modify(remodel_totalothercosts=False)
    assert data['composition']['billSummary']['totalotherCosts'] == {}
```

Still, I think the warning of `because the loop didn't complete` is irrelevant.

**Comment by alirezaaali** [comment]
Good job

**Comment by nedbat** [comment]
Thanks, there does seem to be a bug here.  Simply changing the last `pass` to `a = 1` also makes the warning go away.

**Comment by nedbat** [comment]
This is enough to reproduce the problem:
```
def wrong_loop(x):
    if x:
        for i in [1, 2]:
            print(i)
    else:
        pass

wrong_loop(1)
```

BTW: do you really mean to have `'totalotherCosts'` and `'totalOtherCosts'` as two distinct keys your data structure? That seems really confusing.

**Comment by MaaniBeigy** [comment]
Hi @nedbat,
Sorry that my code was not in a small reproducible form, I was busy and  a bit frustrated. 
Regarding the class, my task was to refactor and rewrite a non-clean code. In the original program, it remodels the data into `totalotherCosts` and deletes the original `totalOtherCosts` later. Regrettably, I spent an hour to figure out these two have different spelling. 
BTW, thanks for the feedback.

**Comment by nedbat** [comment]
So far, I think this is a problem with Python's tracing of a trailing `pass`.  This code is traced properly on Python 3.10.

**Comment by MaaniBeigy** [comment]
Interesting! Thanks a lot.

**Comment by nedbat** [comment]
I've added tests for this case, but I don't think it will be fixed in Python versions before 3.10.

**Comment by nedbat** [comment]
I also wrote https://bugs.python.org/issue44672 to get a decision about a fix in Python.
